AX: VoiceOver not announcing Entering or Leaving when navigating to or from a ordered list as role=listbox

minorninth opened the pull request:
#### 607b0dd8d379f756d33bd9f1d3d7394517de820c
<pre>
AX: VoiceOver not announcing Entering or Leaving when navigating to or from a ordered list as role=listbox
<a href="https://bugs.webkit.org/show_bug.cgi?id=296210">https://bugs.webkit.org/show_bug.cgi?id=296210</a>
<a href="https://rdar.apple.com/156178283">rdar://156178283</a>

Reviewed by Tyler Wilcock.

According to the spec, role=listbox should only be used with role=option as
children. To avoid breakage when role=listbox was used incorrectly, we mapped
it to AXGroup if there are no options. However, this was probably too extreme
when it was put on a static list like a &lt;ul&gt; or &lt;ol&gt; element with list items
as children. Make sure it maps to a static list role in those cases.

* LayoutTests/accessibility/mac/invalid-list-box-valid-list-expected.txt: Added.
* LayoutTests/accessibility/mac/invalid-list-box-valid-list.html: Added.
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::listBoxInterpretation const):
(WebCore::AXCoreObject::isValidListBox const): Deleted.
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::AXCoreObject::isInvalidListBox const): Deleted.
* Source/WebCore/accessibility/cocoa/AXCoreObjectCocoa.mm:
(WebCore::AXCoreObject::rolePlatformString):
* Source/WebCore/accessibility/mac/AccessibilityObjectMac.mm:
(WebCore::AccessibilityObject::subrolePlatformString const):

Canonical link: <a href="https://commits.webkit.org/297634@main">https://commits.webkit.org/297634@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fb5d9fdfdc90b1f2ba01c58bd059daa998a184ef

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/112382 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/32114 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/22591 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/118461 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/62760 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/114344 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/32766 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/40677 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/85372 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/36075 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/115329 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26150 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/101107 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65803 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/25445 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/19246 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/62306 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/95534 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/19321 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/121787 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/39456 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/29374 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/94179 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/39837 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/97348 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/94005 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24034 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39249 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/17046 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/35506 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/39344 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/44832 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/38979 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/42316 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/40722 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/AXCoreObject.cpp:
```diff
@@ -219,35 +219,43 @@ bool AXCoreObject::isTextControl() const
 }
 
-bool AXCoreObject::isValidListBox() const
+ListBoxInterpretation AXCoreObject::listBoxInterpretation() const
```
I like the way you solved this 🙌

twilco commented on LayoutTests/accessibility/mac/invalid-list-box-valid-list.html:
```diff
@@ -0,0 +1,42 @@
+        output += expect("accessibilityController.accessibleElementById('listbox').role", "'AXRole: AXList'");
+        output += expect("accessibilityController.accessibleElementById('listbox').subrole", "'AXSubrole: '");
+        //output += await expectAsync("accessibilityController.accessibleElementById('listbox').role", "'AXRole: AXList'");
+        //output += await expectAsync("accessibilityController.accessibleElementById('listbox').role", "'AXSubrole:'");
```
Discussed outside of GitHub, but seems like we should figure out how to make these `output += await expectAsync`s work before landing.

minorninth committed:
AX: VoiceOver not announcing Entering or Leaving when navigating to or from a ordered list as role=listbox
https://bugs.webkit.org/show_bug.cgi?id=296210
rdar://156178283

Reviewed by Tyler Wilcock.

According to the spec, role=listbox should only be used with role=option as
children. To avoid breakage when role=listbox was used incorrectly, we mapped
it to AXGroup if there are no options. However, this was probably too extreme
when it was put on a static list like a <ul> or <ol> element with list items
as children. Make sure it maps to a static list role in those cases.

* LayoutTests/accessibility/mac/invalid-list-box-valid-list-expected.txt: Added.
* LayoutTests/accessibility/mac/invalid-list-box-valid-list.html: Added.
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::listBoxInterpretation const):
(WebCore::AXCoreObject::isValidListBox const): Deleted.
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::AXCoreObject::isInvalidListBox const): Deleted.
* Source/WebCore/accessibility/cocoa/AXCoreObjectCocoa.mm:
(WebCore::AXCoreObject::rolePlatformString):
* Source/WebCore/accessibility/mac/AccessibilityObjectMac.mm:
(WebCore::AccessibilityObject::subrolePlatformString const):

Canonical link: https://commits.webkit.org/297634@main
